### ISSUE TITLE
Feature Request: Bootstrapping from ISO files

### ISSUE THREAD

**Issue by manuelschwarze** [apptainer/singularity#1033]
## Version of Singularity:
currently using version 2.2.1

### Expected behavior
It would be great to be able to bootstrap containers also based directly on ISO files (URL or local file path) instead of specifying a URL or local folder that contains the unpacked OS install

### Actual behavior
From what I saw in the documentation, Singularity in the moment does not directly support ISO files. I would need to download it, unpack or mount it somehow and then run Singularity based on the local result folder.

### Steps to reproduce behavior
N/A

### Background for this request
When designing processes around building containers, I would like to control what exactly goes into a container. URLs for bootstrapping operating systems may change over time. A simple idea would be to download a base OS as ISO image and conserve it internally as a single file, which will be used for a longer period of time as base OS for all containers that we will build. If Singularity would support bootstrapping based on such an ISO image, similar as it is done already for a docker image, it would simplify the container setup. 
If Docker is allowed and used in a company I would probably go with some official Docker image of the OS, but usually when companies start looking into Singularity one reason is that Docker cannot be used due to security concerns, and in such a case I probably don't want to use a Docker image.

**Comment by vsoch** [comment]
This is a cool idea! I'll give it a shot. Note that we probably won't get it into 2.4 release (impending!) but it could be in the development branch soon after that.

**Comment by GodloveD** [comment]
@manuelschwarze It looks like you could accomplish what you are trying to do by using this in your Singularity recipe: 
```
Bootstrap: localimage
From: /path/to/my/base/container
```
i.e. freeze the OS version locally so that URIs etc do not change and then build new containers from the frozen OS.  Is this correct?  

You can do this already in development (soon to be 2.4).  Does it have to be an ISO?

**Comment by vsoch** [comment]
@GodloveD the extraction of the content doesn't seem to be a big deal, but an iso file has a (not the base of a file system there). Don't we need to do something special with it?

```
ls $SINGULARITY_ROOTFS 
boot  [BOOT]  casper  dists  EFI  install  isolinux  md5sum.txt  pics  pool  preseed  README.diskdefines
```
I don't think it would work as expected if we just extracted into an empty image, but maybe you already wrote the bits to handle the structure above!

**Comment by GodloveD** [comment]
This is Ubuntu right?  Would we need to call `debootstrap` on the contents of the image to build the file system from the local extracted "mirror"?  If so, we are going to need a different bootstrap agent (`yum`, `zypper`, etc) for every flavor of ISO-fied OS.  Not very flexible.

**Comment by vsoch** [comment]
haha yeah that's what I was writing :)

**Comment by cclerget** [comment]
@GodloveD @vsoch by adding iso:// token as MirrorUrl for yum/debootstrap/zypper bootstrap and handle iso accordingly with bootstrap method ?

**Comment by jmstover** [comment]
@cclerget Is iso:// valid? I though only http(s):// and file:// were... 

If you had an iso, you pretty much needed to mount it loopback, then set the file:// URI to that path.

Or are you meaning an iso:// URI in Singularity for handling?

**Comment by cclerget** [comment]
@jmstover handled in Singularity, rather than loop mount, tools like 7z, bsdtar can extract iso content into directory (sure it's longer) but won't require privileges to bootstrap image

**Comment by vsoch** [comment]
that's exactly what I started doing! with @cclerget wisdom, of course. We could easily have an iso:// "bootstrap" that just does checks for 7z first, and then moves forward with similar pathways. We should probably have some of a reorganization of the current scripts - it's a bit jumpy from those in libexec/cli and then bootstrap-scripts, and then image-handlers.

**Comment by manuelschwarze** [comment]
@GodloveD I was hoping to use a direct download of an ISO image that I would put in a local repository without even touching it before building a container, e.g. for CentOS http://isoredirect.centos.org/centos/7/isos/x86_64/CentOS-7-x86_64-Minimal-1708.iso ... I have not tried it, but if I would mount that iso image in linux to a path /centos.iso, I would assume that I could build the Singularity container from it using that mount folder as file:///centos.iso instead of specifying the http:// URL to the unpacked web version of the OS, as done in the documentation example. If that is true, then it should indeed be quite simple to do that mounting as part of the internal Singularity action of bootstrapping, wouldn't it? But maybe I am wrong, and it much more complicated than I thought...

**Comment by GodloveD** [comment]
The problem is that the ISO does not contain an OS file system.  The ISO contains instructions for how to build an OS file system.  So there is another step that has to be carried out after the mount before you have a usable base.  

With Singularity you could download an official version of CentOS from the official Docker Hub registry like so:
```
sudo singularity build centos-base.simg docker://centos:7
```
Then `centos-base.simg` would be a local file containing a pristine CentOS file system.  That's what you want right?

Then when you are ready to build a second container using the pristine and reproducible `centos-base.simg` as your base you would write a recipe like so:
```
Bootstrap: localimage
From: centos-base.simg

%post
    yum -y update
    yum -y install vim
```
Or some such thing.  Save the recipe with the name `Singularity`.  Then you would build the second container like this:
```
sudo singularity build vim.simg Singularity
```
And you could go on building containers from this reproducible base for as long as you like.

**Comment by manuelschwarze** [comment]
@GodloveD What you describe as alternative to ISO makes absolutely sense. It was not possible in version 2.2.1 which I used when I reached out to you. But I tried it with version 2.4 successfully. This is for my purposes even better, because deriving multiple container flavors from a single base OS container (that I can also archive) is performing very fast.

I was under the impression that an ISO image has the same information that your code is retrieving today from the OS URL that one provides. Apparently wrong. In that case I would withdraw my feature request :-) ... unless you think it may be worth looking into this for another reason. Feel free to close it.
